packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -367,130 +218,98 @@ library ExchangeBlocks
-        if (numConditionalTransfers > 0) {
-            require(S.onchainDataAvailability, "CONDITIONAL_TRANSFERS_REQUIRE_OCDA");
+        uint numConditionalTransactions = data.toUint32(offset);
+        offset += 4;
 
-            require(auxiliaryData.length == numConditionalTransfers * 4, "INVALID_AUXILIARYDATA_LENGTH");
-            offset += 4;
+        if (numConditionalTransactions > 0) {
+            require(S.rollupEnabled, "CONDITIONAL_TRANSACTIONS_AVAILABLE_ONLY_IN_ROLLUP_MODE");
 
-            uint24 operatorAccountID = data.bytesToUint24(offset);
-            offset += 3;
+            // Cache the domain seperator to save on SLOADs each time it is accessed.
+            ExchangeData.BlockContext memory ctx = ExchangeData.BlockContext({
+                DOMAIN_SEPARATOR: S.DOMAIN_SEPARATOR
+            });
 
-            // Run over all conditional transfers
-            uint previousTransferOffset = 0;
-            for (uint i = 0; i < auxiliaryData.length; i += 4) {
-                // auxiliaryData contains the transfer index (4 bytes) for each conditional transfer
-                // in ascending order.
-                uint transferOffset = offset + auxiliaryData.bytesToUint32(i) * 15 + 15;
-                require(transferOffset > previousTransferOffset, "INVALID_AUXILIARYDATA_DATA");
-
-                // Get the transfer data
-                uint transferData;
-                assembly {
-                    transferData := and(mload(add(data, transferOffset)), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
-                }
+            ExchangeData.AuxiliaryData[] memory txAuxiliaryData = abi.decode(auxiliaryData, (ExchangeData.AuxiliaryData[]));
+            require(txAuxiliaryData.length == numConditionalTransactions, "AUXILIARYDATA_INVALID_LENGTH");
 
-                // Check that this is a conditional transfer
-                uint transferType = (transferData >> 112) & 0xFF;
-                require(transferType == 1, "INVALID_AUXILIARYDATA_DATA");
+            // uint24 operatorAccountID = data.toUint24(offset);
+            offset += 3;
 
-                // Update the onchain state
-                consumeConditionalTransfer(
-                    S,
-                    operatorAccountID,
-                    transferData
-                );
+            // Run over all conditional transactions
+            uint previousTransactionOffset = 0;
+            for (uint i = 0; i < txAuxiliaryData.length; i++) {
+                // Each conditional transaction needs to be processed from left to right
+                uint transactionOffset = offset + txAuxiliaryData[i].txIndex * ExchangeData.TX_DATA_AVAILABILITY_SIZE();
+                require(transactionOffset > previousTransactionOffset, "AUXILIARYDATA_INVALID_ORDER");
+
+                // Get the transaction data
+                bytes memory txData = data.slice(transactionOffset, ExchangeData.TX_DATA_AVAILABILITY_SIZE());
```

Comment:
Could be, but I haven't measured it, but the slice is actually very efficient. I actually tested this out in the signature utils because it's used there as well and saw almost no gas improvement at all. In any case, it's probably very low so probably comes down to what makes the code better.

Reply:
Measured the gas use to be 185 gas for the `slice`. A simple `require` check costs 35 gas.

So not too expensive, but on the other it doesn't really change much in the code, just add the offset into the `process` function and it works the same. The only extra danger I can think of is that function could read across the transaction boundary, but then the implementation would be bad any way. So we may as well do it. :)